### ISSUE TITLE
v9 useTransition regressions

### ISSUE THREAD

**Issue by aleclarson** [react-spring/react-spring#676]
noticed a useTransition regression, my modal close animation stopped working after upgrading (it's now opening with animation and closing instantly on v9)

https://github.com/devhubapp/devhub/blob/6aa24bb96658ce2541f69f9eccfc3d65826c726b/packages/components/src/components/modals/ModalRenderer.tsx#L130-L172

_Originally posted by @brunolemos in https://github.com/react-spring/react-spring/issues/642#issuecomment-485619594_

**Comment by aleclarson** [comment]
There seem to be regression/change of behavior with `useTransition` where values are cached by key. This is problematic for us because we use this mechanism to skip transitions by using the same key, but still need the new object to be propagated to the view.

Sample:

react-spring 8 - https://codesandbox.io/s/ly068v5x2q
react-spring 9 - https://codesandbox.io/s/434popoxyw

In both cases no animation is expected (correct as the key is always the same), however in v8 `item` holds newly propagated value whereas in v9 `item` holds value from previous render.


Is this intended change?

_Originally posted by @Kukkimonsuta in https://github.com/react-spring/react-spring/issues/642#issuecomment-491529463_

**Comment by aleclarson** [comment]
Similar issue reported by @rafalfilipek on Twitter:
https://twitter.com/rafalfilipek/status/1146122279836880896

**Comment by m-rutter** [comment]
I just hit this issue while using the v9 beta-9. 

Below is a fairly minimal example. In latest v8 the content will update after clicking the button. In v9 the content will not update because the `item` in the transition collection is a cached object from the first render. 

I'm guessing its necessary to cache the items by key for when an item leaves, but I'm guessing that the cache doesn't get updated again after the first render.

```tsx
import React, {useState} from 'react'
import { useTransition, animated} from "react-spring";

function Test() {
    const [items, setItems] = useState([{ id: 1, content: "some content" }]);

    console.log(items[0].content); // changes after pressing button

    const transitions = useTransition(items, (item) => item.id, {
        enter: { opacity: 1 },
        leave: { opacity: 0 },
        initial: { opacity: 0 }
    });

    const updateContent = () => {
        setItems([
            {
                id: 1,
                content: "some content, with even more content than before"
            }
        ]);
    };

    return (
        <>
            <button onClick={updateContent}>change content</button>
            {transitions.map(({ item, key, props }) => (
                <animated.div key={key} style={props}>
                    {item.content}
                </animated.div>
            ))}
        </>
    );
}


```

**Comment by aleclarson** [comment]
@m-rutter The item caching issue will be fixed when #750 is merged 👍

**Comment by brunolemos** [comment]
> noticed a useTransition regression, my modal close animation stopped working after upgrading (it's now opening with animation and closing instantly on v9)

I've found that changing from `reset: true` to `reset: false` fixes this. Maybe this is the intended behavior and v8 is actually the one behaving incorrectly? 

Nevertheless, I still can't upgrade to v9 because of other bugs. I might open issues for them.

**Comment by aleclarson** [comment]
Now available in v9.0.0-rc.2 #985
